### PR TITLE
[RLMObjectBase] Don't crash when calling isEqual: against a non-RLMObject...

### LINKED ISSUE #1547: isEqual: between Realm object and non-Realm object causes a crash
Looks like `[realmObject isEqual:nonRealmObject]` causes a crash — unless `nonRealmObject` implements, for example,
`- (id)realm { return nil; }`

### PR DESCRIPTION
**Pull request by segiddins**
Closes #1547.

\c @alazier @tgoyne

### CODE REVIEW

File: Realm/RLMObjectBase.mm [hunk 1]
```diff
@@ -256,12 +256,12 @@ - (BOOL)isEqualToObject:(RLMObject *)object {
...
-        return [self isEqualToObject:object];
-    }
-    else {
-        return [super isEqual:object];
+    if (RLMObjectBase *other = RLMDynamicCast<RLMObjectBase>(object)) {
+        if (_objectSchema.primaryKeyProperty) {
+            return [self isEqualToObject:other];
+        }
     }
+    return [super isEqual:object];
```
Comment (alazier) [inline Realm/RLMObjectBase.mm]: Wont this always be NO?

**Comment by alazier** [comment]
should definitely test this.

File: Realm/RLMObjectBase.mm (hunk 1 above)
Comment (segiddins) [inline Realm/RLMObjectBase.mm]: no, it'll be `YES` if and only if `self == object` (because that's the `NSObject` implementation)

Comment (tgoyne) [inline Realm/RLMObjectBase.mm]: i.e. it'll always be NO because we would have already returned YES if that was the case.

Comment (segiddins) [inline Realm/RLMObjectBase.mm]: not necessarily, since we only check `isEqualToObject:` when there's a primary key property

Comment (tgoyne) [inline Realm/RLMObjectBase.mm]: Oh, I misread it and missed that the `self == object` check a few lines up wasn't part of this method...

**Comment by jpsim** [comment]
I've added tests and a changelog entry for this. Anything else? /cc @segiddins @tgoyne

**Comment by segiddins** [comment]
:+1:

**Pull request merged**